examples/templates/java/kafka-to-pubsub/src/test/java/org/apache/beam/templates/KafkaToPubsubTest.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.templates;
+
+import static org.apache.beam.templates.KafkaPubsubConstants.PASSWORD;
+import static org.apache.beam.templates.KafkaPubsubConstants.USERNAME;
+import static org.apache.beam.templates.kafka.consumer.Utils.getKafkaCredentialsFromVault;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.beam.templates.kafka.consumer.Utils;
+import org.apache.kafka.common.config.SaslConfigs;
+import org.apache.kafka.common.security.scram.ScramMechanism;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Test of KafkaToPubsub. */
+@RunWith(JUnit4.class)
+public class KafkaToPubsubTest {
```

Comment:
Maybe I am missing something here but these tests only validate the config.
Is it possible to have a more e2e test where we create & execute a pipeline like what we had before ?

Reply:
We've made an investigation towards e2e tests and for such tests we need Kafka and Pub/Sub being set up in some environment or in a GCP account with billing. That is why seems like there is no way to do it properly.

Reply:
You might take a look at how our tests for PubsubIO and KafkaIO work as they have to contend with this same problem.

For KafkaIO, we use [testcontainers](https://github.com/apache/beam/blob/02da3ae62e9025ff093747bb7a098982d34eff1b/sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java#L336) to stand up a fake Kafka service to test against.

For E2E tests of PubsubIO and other GCP-specific features we have the apache-beam-testing project. We could probably give you access to this. Alternatively, there is now a [PubSub emulator](https://cloud.google.com/pubsub/docs/emulator) that you could use to stand up a fake PubSub service, just like the Kafka testcontainer. I don't think there are any examples using it in Beam yet though.

Reply:
Thank you for bringing up the requirement of the e2e tests, we understand the solution needs it.
Our focus is to get this example available to the community and be able to collect any feedback from customers using this example in their scenarios and the approach to the examples. We propose to keep this PR scope, create a JIRA ticket (for [Kafka](https://issues.apache.org/jira/browse/BEAM-11410) and [Pub/Sub](https://issues.apache.org/jira/browse/BEAM-11411)) to extend this example with e2e tests, and set customer expectations on e2e testing in the [readme](https://github.com/apache/beam/pull/13112/commits/e8ca92d68646cec459d3ecd94b0efe7551699268) for this PR.

Reply:
I'm a +1 for @ilya-kozyrev's proposal. 
If the solution passes all tests, I'm supportive of merging the solution so we entice users to test it according to their use cases. The consideration here is to complete the two steps he suggests: adding the note on e2e tests needed and creating Jira tickets to create these.

Reply:
Sounds good. Please add the note and the link the JIRA tickets in that case.